File: server/internal/admin/db/plaid_credentials.go
package db

import (
	"context"
	"database/sql"
	"errors"
	"strings"

	"github.com/samber/lo"

	"tallyo/internal/apierror"
	"tallyo/internal/clients"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
)

func (s *Store) CreatePlaidCredential(ctx context.Context, input model.CreatePlaidCredentialInput) (*model.PlaidCredential, error) {
	clientID := strings.TrimSpace(input.ClientID)
	secret := strings.TrimSpace(input.Secret)
	if clientID == "" {
		return nil, apierror.Publicf("client_id is required")
	}
	if secret == "" {
		return nil, apierror.Publicf("secret is required")
	}
	label := strings.TrimSpace(lo.FromPtr(input.Label))
	id, err := s.q.CreatePlaidCredential(ctx, dbgen.CreatePlaidCredentialParams{ClientID: clientID, Secret: secret, Environment: strings.ToLower(string(input.Environment)), Label: lo.EmptyableToPtr(label)})
	if err != nil {
		return nil, err
	}
	return s.PlaidCredential(ctx, int32(id))
}

func (s *Store) UpdatePlaidCredential(ctx context.Context, input model.UpdatePlaidCredentialInput) (*model.PlaidCredential, error) {
	secret := strings.TrimSpace(input.Secret)
	if secret == "" {
		return nil, apierror.Publicf("secret is required")
	}
	rows, err := s.q.UpdatePlaidCredential(ctx, dbgen.UpdatePlaidCredentialParams{Secret: secret, Environment: strings.ToLower(string(input.Environment)), ID: int64(input.ID)})
	if err != nil {
		return nil, err
	}
	if rows == 0 {
		return nil, sql.ErrNoRows
	}
	return s.PlaidCredential(ctx, int32(input.ID))
}

func (s *Store) DeletePlaidCredential(ctx context.Context, id int32) (bool, error) {
	rows, err := s.q.DeletePlaidCredential(ctx, dbgen.DeletePlaidCredentialParams{ID: int64(id)})
	return rows > 0, err
}

func (s *Store) PlaidCredentials(ctx context.Context) ([]*model.PlaidCredential, error) {
	rows, err := s.q.ListPlaidCredentials(ctx, dbgen.ListPlaidCredentialsParams{})
	return dbutil.MapRows(rows, err, plaidCredentialFromRow)
}

func (s *Store) PlaidCredential(ctx context.Context, id int32) (*model.PlaidCredential, error) {
	credentials, err := s.PlaidCredentialsByIDs(ctx, []int32{id})
	return dbutil.MapSingle(credentials, err, id)
}

func (s *Store) PlaidCredentialsByIDs(ctx context.Context, ids []int32) (map[int32]*model.PlaidCredential, error) {
	if len(ids) == 0 {
		return map[int32]*model.PlaidCredential{}, nil
	}
	credentialIDs := u.Map(ids, func(id int32) int64 { return int64(id) })
	rows, err := s.q.ListPlaidCredentials(ctx, dbgen.ListPlaidCredentialsParams{Ids: credentialIDs})
	toCredentialByID := func(row dbgen.ListPlaidCredentialsRow) (int32, *model.PlaidCredential) {
		return int32(row.ID), plaidCredentialFromRow(row)
	}
	return dbutil.AssociateRows(rows, err, toCredentialByID)
}

func plaidCredentialFromRow(row dbgen.ListPlaidCredentialsRow) *model.PlaidCredential {
	return &model.PlaidCredential{ID: int32(row.ID), ClientID: row.ClientID, Environment: model.PlaidEnvironment(strings.ToUpper(row.Environment)), Label: row.Label, ItemCount: int32(row.ItemCount), CreatedAt: row.CreatedAt}
}

func (s *Store) CredentialByID(ctx context.Context, id int32) (*clients.PlaidCredential, error) {
	row, err := s.q.PlaidCredentialSecretByID(ctx, dbgen.PlaidCredentialSecretByIDParams{ID: int64(id)})
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &clients.PlaidCredential{ClientID: row.ClientID, Secret: row.Secret, Environment: row.Environment}, nil
}

File: server-rs/src/admin/store/plaid_credentials.rs
use std::collections::HashMap;

use anyhow::{Context, Result, ensure};
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    admin::PlaidCredential,
    apierror::ApiError,
    database::queries,
    schema::{CreatePlaidCredentialInput, UpdatePlaidCredentialInput},
};

const CLIENT_ID_REQUIRED: &str = "client_id is required";
const SECRET_REQUIRED: &str = "secret is required";

pub async fn create_plaid_credential(pool: &SqlitePool, input: CreatePlaidCredentialInput) -> Result<PlaidCredential> {
    let client_id = input.client_id.trim();
    let secret = input.secret.trim();
    ensure!(!client_id.is_empty(), ApiError::bad_input(CLIENT_ID_REQUIRED));
    ensure!(!secret.is_empty(), ApiError::bad_input(SECRET_REQUIRED));
    let label = input.label.as_deref().map(str::trim).filter(|label| !label.is_empty());
    let environment = input.environment.to_string().to_lowercase();
    let id = queries::create_plaid_credential(
        pool,
        queries::CreatePlaidCredentialParams {
            client_id,
            secret,
            environment: &environment,
            label,
        },
    )
    .await?
    .id;
    plaid_credential(pool, id)
        .await?
        .context("find created plaid credential")
}

pub async fn update_plaid_credential(pool: &SqlitePool, input: UpdatePlaidCredentialInput) -> Result<PlaidCredential> {
    let secret = input.secret.trim();
    ensure!(!secret.is_empty(), ApiError::bad_input(SECRET_REQUIRED));
    let environment = input.environment.to_string().to_lowercase();
    let updated = queries::update_plaid_credential(
        pool,
        queries::UpdatePlaidCredentialParams {
            secret,
            environment: &environment,
            id: i64::from(input.id),
        },
    )
    .await?;
    ensure!(updated > 0, "plaid credential not found");
    plaid_credential(pool, i64::from(input.id))
        .await?
        .context("find updated plaid credential")
}

pub async fn delete_plaid_credential(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<bool> {
    queries::delete_plaid_credential(executor, queries::DeletePlaidCredentialParams { id })
        .await
        .map(|rows| rows > 0)
        .map_err(Into::into)
}

pub async fn plaid_credentials(executor: impl Executor<'_, Database = Sqlite>) -> Result<Vec<PlaidCredential>> {
    queries::list_plaid_credentials(executor, queries::ListPlaidCredentialsParams::default())
        .await
        .map_err(Into::into)
        .and_then(|rows| rows.into_iter().map(PlaidCredential::try_from).collect())
}

pub async fn plaid_credential(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
) -> Result<Option<PlaidCredential>> {
    plaid_credentials_by_ids(executor, &[id])
        .await
        .map(|mut credentials| credentials.remove(&id))
}

pub async fn plaid_credentials_by_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    ids: &[i64],
) -> Result<HashMap<i64, PlaidCredential>> {
    if ids.is_empty() {
        return Ok(HashMap::new());
    }
    queries::list_plaid_credentials(executor, queries::ListPlaidCredentialsParams { ids: Some(ids) })
        .await
        .map_err(Into::into)
        .and_then(|rows| {
            rows.into_iter()
                .map(|row| {
                    let id = row.id;
                    PlaidCredential::try_from(row).map(|credential| (id, credential))
                })
                .collect::<Result<HashMap<_, _>>>()
        })
}

pub async fn plaid_credential_secret_by_id(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
) -> Result<Option<crate::clients::plaid::PlaidCredential>> {
    queries::plaid_credential_secret_by_id_opt(executor, queries::PlaidCredentialSecretByIdParams { id })
        .await
        .map(|row| {
            row.map(|row| crate::clients::plaid::PlaidCredential {
                client_id: row.client_id,
                secret: row.secret,
                environment: row.environment.as_str().into(),
            })
        })
        .map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use super::{
        CLIENT_ID_REQUIRED, SECRET_REQUIRED, create_plaid_credential, delete_plaid_credential, plaid_credential,
        plaid_credential_secret_by_id, plaid_credentials, plaid_credentials_by_ids, update_plaid_credential,
    };
    use crate::{
        database::{dbtest, queries},
        schema::{CreatePlaidCredentialInput, PlaidEnvironment, UpdatePlaidCredentialInput},
        testutil::store::{create_owner, seed_plaid_item},
    };

    fn create_input() -> CreatePlaidCredentialInput {
        CreatePlaidCredentialInput {
            client_id: " client-1 ".to_owned(),
            secret: " secret-1 ".to_owned(),
            environment: PlaidEnvironment::Development,
            label: Some(" Primary ".to_owned()),
        }
    }

    #[tokio::test]
    async fn validates_creates_updates_and_deletes_credentials() {
        let pool = dbtest::open().await.unwrap();
        let mut missing_client_id = create_input();
        missing_client_id.client_id = " ".to_owned();
        assert_eq!(
            create_plaid_credential(&pool, missing_client_id)
                .await
                .unwrap_err()
                .to_string(),
            CLIENT_ID_REQUIRED
        );
        let mut missing_secret = create_input();
        missing_secret.secret = " ".to_owned();
        assert_eq!(
            create_plaid_credential(&pool, missing_secret)
                .await
                .unwrap_err()
                .to_string(),
            SECRET_REQUIRED
        );

        let created = create_plaid_credential(&pool, create_input()).await.unwrap();
        assert_eq!(created.client_id, "client-1");
        assert_eq!(created.label.as_deref(), Some("Primary"));
        assert_eq!(created.environment, PlaidEnvironment::Development);
        assert_eq!(plaid_credentials(&pool).await.unwrap(), vec![created.clone()]);
        assert_eq!(
            plaid_credential_secret_by_id(&pool, i64::from(created.id))
                .await
                .unwrap()
                .unwrap()
                .secret,
            "secret-1"
        );

        let updated = update_plaid_credential(
            &pool,
            UpdatePlaidCredentialInput {
                id: created.id,
                secret: "secret-2".to_owned(),
                environment: PlaidEnvironment::Production,
            },
        )
        .await
        .unwrap();
        assert_eq!(updated.environment, PlaidEnvironment::Production);
        assert_eq!(plaid_credentials_by_ids(&pool, &[]).await.unwrap().len(), 0);
        assert!(delete_plaid_credential(&pool, i64::from(created.id)).await.unwrap());
        assert!(plaid_credential(&pool, i64::from(created.id)).await.unwrap().is_none());
        assert!(!delete_plaid_credential(&pool, i64::from(created.id)).await.unwrap());
    }

    #[tokio::test]
    async fn counts_active_items_and_rejects_invalid_storage_environments() {
        let pool = dbtest::open().await.unwrap();
        let owner = create_owner(&pool, "Owner").await.unwrap();
        let (item_id, _) = seed_plaid_item(&pool, &owner, "item").await.unwrap();
        let credential_id = i64::from(plaid_credentials(&pool).await.unwrap().pop().unwrap().id);

        assert_eq!(
            plaid_credentials_by_ids(&pool, &[credential_id]).await.unwrap()[&credential_id].item_count,
            1
        );
        assert!(delete_plaid_credential(&pool, credential_id).await.is_err());

        queries::delete_plaid_item(&pool, queries::DeletePlaidItemParams { id: item_id })
            .await
            .unwrap();
        assert!(delete_plaid_credential(&pool, credential_id).await.unwrap());

        let invalid = create_plaid_credential(&pool, create_input()).await.unwrap();
        sqlx::query("UPDATE plaid_credentials SET environment = 'invalid' WHERE id = ?")
            .bind(invalid.id)
            .execute(&pool)
            .await
            .unwrap();
        assert_eq!(
            plaid_credentials(&pool).await.unwrap_err().to_string(),
            "parse persisted Plaid environment \"invalid\""
        );
    }
}
